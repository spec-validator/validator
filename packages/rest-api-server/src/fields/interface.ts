/*
import { Field, SpecUnion, isArraySpec, isFieldSpec, isObjectSpec } from './core'
import { undefinedField, arrayField, objectField } from './fields'

export const getFieldForSpec = <DeserializedType> (
  spec: SpecUnion<DeserializedType>,
  allowExtraFields = false
): Field<DeserializedType> => {
  if (isFieldSpec(spec)) {
    return spec
  } else if (isArraySpec(spec)) {
    return arrayField(getFieldForSpec(spec[0])) as unknown as Field<DeserializedType>
  } else if (isObjectSpec(spec)) {
    return objectField(Object.fromEntries(
      Object.entries(spec).map(([key, value]) => [key, getFieldForSpec(value)])
    ), allowExtraFields) as unknown as Field<DeserializedType>
  } else {
    return undefinedField()
  }
}
*/
