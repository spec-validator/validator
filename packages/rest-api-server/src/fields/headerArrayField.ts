import { declareField, withErrorDecoration } from '@spec-validator/validator/core'
import { FieldWithRegExpSupport } from '@spec-validator/validator/fields/segmentField'

export interface HeaderArrayField<T> extends FieldWithRegExpSupport<T[]> {
  readonly itemField: FieldWithRegExpSupport<T>,
  readonly regex: RegExp,
  serialize(input: T[]): string
}

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Content-Language
const SEPARATOR = ', '

export default declareField('@spec-validator/rest-api/fields/headerArrayField', <T>(
  itemField: FieldWithRegExpSupport<T>
): HeaderArrayField<T> => {

  const fieldWithStringSupport = itemField.getStringField()
  const regex = fieldWithStringSupport.regex.source

  const result = {
    itemField,
    regex: new RegExp(`${regex}(${SEPARATOR}${regex})+`),
    validate: (value: any): T[] => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      return value.split(SEPARATOR).map(
        (it, index) => withErrorDecoration(index, () => fieldWithStringSupport.validate(decodeURI(it)))
      )
    },
    serialize: (deserialized: T[]): string =>
      deserialized.map(
        (it, index) => withErrorDecoration(index, () => encodeURI(fieldWithStringSupport.serialize(it)))
      ).join(SEPARATOR),
  } as HeaderArrayField<T>

  result.getStringField = () => result

  return result
})
