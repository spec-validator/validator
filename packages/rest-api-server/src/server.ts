import http from 'http'
import { URL } from 'url'

import createHandler, { getServerConfigs, ServerConfig, WildCardRequest, WildCardResponse } from './handler'

const SUPPORTED_PROTOCOLS= {
  'http': 80,
  'https': 443,
} as const

const supportedProtocols = new Set(Object.keys(SUPPORTED_PROTOCOLS))

const verifyProtocol = (protocol: string): protocol is keyof typeof SUPPORTED_PROTOCOLS =>
  supportedProtocols.has(protocol)

const getPort = (baseUrl: string): number => {
  const url = new URL(baseUrl)
  // Node's protocol has a trailing column
  const protocol = url.protocol.replace(/:$/, '')
  if (!verifyProtocol(protocol)) {
    throw `Protocol ${url.protocol} is not supported`
  }
  return url.port ? Number.parseInt(url.port) : SUPPORTED_PROTOCOLS[protocol]
}

type ConfiguredServer = http.Server & { serve: () => http.Server }

const getData = async (msg: http.IncomingMessage): Promise<string> => new Promise<string> ((resolve, reject) => {
  const chunks: string[] = []
  msg.on('readable', () => chunks.push(msg.read()))
  msg.on('error', reject)
  msg.on('end', () => resolve(chunks.join('')))
})

const getWildcardRequest = async (
  request: http.IncomingMessage,
): Promise<WildCardRequest> => ({
  method: request.method as string,
  url: request.url as string,
  body: await getData(request),
  headers: request.headers,
})

const toHttpResponse = async (
  config: ServerConfig,
  resp: WildCardResponse,
  response: http.ServerResponse
): Promise<void> => {
  Object.entries(resp.headers || {}).forEach(([key, value]) => {
    response.setHeader(key, value as any)
  })

  response.statusCode = resp.statusCode

  if (resp.body !== undefined) {
    response.write(
      resp.body,
      config.encoding
    )
  }

  response.end()
}

// eslint-disable-next-line max-statements
const proc =  (
  config: ServerConfig
): ((request: http.IncomingMessage, response: http.ServerResponse) => Promise<void>) => {
  const handleWildCardRounte = createHandler(config)
  return async (request: http.IncomingMessage, response: http.ServerResponse): Promise<void> => {
    const wildCardRequest = await getWildcardRequest(request)
    const wildCardResponse = await handleWildCardRounte(wildCardRequest)
    await toHttpResponse(config, wildCardResponse, response)
  }
}

export const createServer = (
  config: Partial<ServerConfig>,
): ConfiguredServer => {
  const merged = getServerConfigs(config)
  const server = http.createServer(proc(merged)) as ConfiguredServer
  server.serve = () => server.listen(getPort(merged.baseUrl))
  return server
}
