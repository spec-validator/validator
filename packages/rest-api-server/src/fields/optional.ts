import { Any, Optional } from '@spec-validator/utils/util-types'
import { Field, declareField } from '@spec-validator/validator/core'
import { FieldWithStringInputSupport } from '@spec-validator/validator/fields/segmentField'

export interface OptionalHeaderField<T extends Any> extends Field<Optional<T>> {
  readonly innerField: FieldWithStringInputSupport<T>
}

export default declareField('@spec-validator/rest-api/fields/optional', <T extends Any> (
  innerField: FieldWithStringInputSupport<T>
): OptionalHeaderField<T> => {
  const realField = innerField.getFieldWithRegExp()
  return {
    innerField,
    validate: (value: any): Optional<T> => {
      if (value === undefined) {
        return value
      }
      return realField.validate(value)
    },
    serialize: (deserialized: Optional<T>) =>
      deserialized === undefined ? deserialized : realField.serialize(deserialized) as any,
  }
})
