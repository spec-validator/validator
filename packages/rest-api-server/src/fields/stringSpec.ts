import { Any } from '@spec-validator/utils/util-types'
import { FieldWithStringInputSupport } from '@spec-validator/validator/fields/segmentField'

export type StringObjectSpec<DeserializedType extends Record<string, Any> = Record<string, Any>> = {
  [P in keyof DeserializedType]: FieldWithStringInputSupport<DeserializedType[P]>
}

export type StringArraySpec<DeserializedType extends Any[] = Any[]> = StringSpecUnion<DeserializedType[number]>[]

export type StringSpecUnion<DeserializedType> =
  StringObjectSpec | StringArraySpec | FieldWithStringInputSupport<DeserializedType> | undefined
