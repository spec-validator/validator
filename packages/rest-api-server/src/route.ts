import { TypeHint } from '@spec-validator/validator'
import { Field, isFieldSpec, ObjectSpec, SpecUnion, OfType } from '@spec-validator/validator/core'
import { $, unionField } from '@spec-validator/validator/fields'
import { ConstantField } from '@spec-validator/validator/fields/constantField'
import { UnionField } from '@spec-validator/validator/fields/unionField'
import { Any, WithoutOptional } from '@spec-validator/utils/util-types'

export type StringMapping = Record<string, Any>

export type HeaderMapping = Record<string, string | string[] | number>

export type DataMapping = StringMapping | Any

export type RequestSpec<
  Method extends string = string,
  PathParams extends StringMapping | unknown = StringMapping | unknown,
  Body extends Any = Any,
  QueryParams extends StringMapping = StringMapping,
  Headers extends HeaderMapping = HeaderMapping,
> = {
  readonly method: ConstantField<Method>,
  readonly pathParams: typeof $ & Field<PathParams>,
  readonly body?: SpecUnion<Body>,
  readonly headers?: ObjectSpec<Headers>,
  readonly queryParams?: ObjectSpec<QueryParams>
}

export type ResponseSpec<
  StatusCode extends number = number,
  Body extends Any = Any,
  Headers extends HeaderMapping = HeaderMapping,
> = {
  readonly statusCode: ConstantField<StatusCode>,
  readonly body?: SpecUnion<Body>,
  readonly headers?: ObjectSpec<Headers>,
}

type ResponseField<Spec extends ResponseSpec=ResponseSpec > = WithoutOptional<Spec>

// TODO: how to extract schema from
type ResponsesSpec<ResponseVariants extends ResponseField[] = ResponseField[]> =
  UnionField<ResponseVariants>

export type Route<
  ReqSpec extends RequestSpec = RequestSpec,
  RespSpec extends ResponsesSpec | ResponseSpec = ResponsesSpec | ResponseSpec
> = {
  readonly request: ReqSpec,
  readonly response: RespSpec,
  readonly handler: (request: WithoutOptional<TypeHint<ReqSpec>>) => Promise<
    WithoutOptional<TypeHint<RespSpec>>
  >
}

export const isResponsesSpec = (spec: ResponsesSpec | ResponseSpec): spec is ResponsesSpec =>
  isFieldSpec(spec) && (spec as unknown as OfType<string>).type !== unionField.type

export const route = <
  ReqSpec extends RequestSpec = RequestSpec,
  RespSpec extends ResponsesSpec | ResponseSpec = ResponsesSpec | ResponseSpec,
  >(spec: {
    request: ReqSpec,
    response: RespSpec
  }): {
    handler: (handler:
      (request: WithoutOptional<TypeHint<ReqSpec>>) => Promise<WithoutOptional<TypeHint<RespSpec>>>
    ) => Route
  } => ({
    handler: (handler): Route => ({
      ...spec,
      handler: async (request) => await handler(request as any),
    }),
  })
