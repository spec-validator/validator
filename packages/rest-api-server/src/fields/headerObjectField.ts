import { declareField, FieldWithStringSupport, TypeHint, withErrorDecoration } from '@spec-validator/validator/core'

import { StringObjectSpec } from './stringSpec'

export interface HeaderObjectField<Spec extends StringObjectSpec = StringObjectSpec>
extends FieldWithStringSupport<TypeHint<Spec>> {
  readonly objectSpec: Spec,
  readonly separator: string
  readonly regex: RegExp
  serialize(input: TypeHint<Spec>): string
}

// https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cookie
const SEPARATOR = '; '

export default declareField('@spec-validator/rest-api/fields/headerObjectField', <Spec extends StringObjectSpec>(
  objectSpec: Spec
): HeaderObjectField<Spec> => {

  const regex = '.+=.+'

  const result = {
    objectSpec,
    regex: new RegExp(`${regex}(${SEPARATOR}${regex})*`),
    validate: (value: any): TypeHint<Spec> => {
      if (typeof value !== 'string') {
        throw 'Not a string'
      }
      const payload = Object.fromEntries(value.split(SEPARATOR)
        .map(it =>it.split(/=(.*)/, 2)))

      return Object.fromEntries(Object.entries(objectSpec).map(([key, valueSpec]) => [
        key, withErrorDecoration(key, () => valueSpec.getStringField().validate(
          decodeURI(payload[key]))
        ),
      ])) as TypeHint<Spec>
    },
    serialize: (deserialized: TypeHint<Spec>): string =>
      Object.entries(objectSpec)
        .map(([key, valueSpec]) => [
          key, withErrorDecoration(key, () => encodeURI(
            valueSpec.getStringField().serialize(deserialized[key])
          )),
        ])
        .map(([key, value]) => `${key}=${value}`)
        .join(SEPARATOR),
  } as HeaderObjectField<Spec>

  result.getStringField = () => result

  return result
})
