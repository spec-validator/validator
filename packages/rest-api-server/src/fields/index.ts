import headerArrayField from './headerArrayField'
import headerObjectField from './headerObjectField'

export { headerObjectField, headerArrayField }
