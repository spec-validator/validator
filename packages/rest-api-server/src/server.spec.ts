import request from 'supertest'

import {
  $, constantField, numberField, objectField, optional, stringField, unionField,
} from '@spec-validator/validator/fields'


import { createServer, _ } from './server'
import { route } from './route'
import { headerArrayField, headerObjectField } from './fields'

const itemSpec = {
  title: stringField(),
  description: stringField(),
}

const ofItems = {
  body: [itemSpec],
}

const ofItem = {
  body: itemSpec,
}

// TODO: disablable error on extra keys for object field
const server = createServer({routes: [
  _.GET($._('/expected-error')).spec(
    {
      response: {
        body: constantField(42),
      },
    }
  ).handler(
    async () => {
      throw {
        statusCode: 442,
        isPublic: true,
        reason: 'Boom!',
      }
    }
  ),
  route({
    request: {
      method: constantField('GET'),
      pathParams: $._('/multi-response'),
    },
    response: unionField({
      statusCode: constantField(400),
    }, {
      statusCode: constantField(200),
    }),
  }).handler( async () => ({
    statusCode: 200,
  }),
  ),
  _.GET($._('/unexpected-error')).spec(
    {
      response: {
        body: constantField(42),
      },
    }
  ).handler(
    async () => {
      throw {
        reason: 'Boom!',
      }
    }
  ),
  _.GET($._('/items')).spec(
    {
      response: ofItems,
    },
  ).handler(
    async () => ({
      body: [
        {
          title: 'Item N',
          description: 'Description',
        },
      ],
    })
  ),
  _.POST($._('/items')).spec(
    {
      request: ofItem,
      response: {
        body: numberField(),
        headers: {
          title: stringField(),
        },
      },
    },
  ).handler(
    async () => ({
      body: 42,
      headers: {
        title: 'Foo',
      },
    })
  ),
  _.GET($._('/items/')._('id', numberField())).spec(
    {
      response: ofItem,
    },
  ).handler(
    async (req) => ({
      body:
        {
          title: `Item ${req.pathParams.id}`,
          description: 'Description',
        },
    })
  ),
  _.PUT($._('/items/')._('id', numberField())).spec(
    {
      request: ofItem,
    },
  ).handler(
    async () => undefined
  ),
  _.PATCH($._('/items/')._('id', numberField())).spec(
    {
      request: {
        body: objectField({
          title: optional(stringField()),
          description: optional(stringField()),
        }),
      },
    },
  ).handler(
    async () => undefined
  ),
  _.DELETE($._('/items/')._('id', numberField())).spec({}).handler(
    async () => undefined
  ),
  _.GET($._('/with-complex-headers')).spec(
    {
      response: {
        headers: {
          one: stringField(),
          many: headerArrayField(numberField()),
          cookies: headerObjectField({
            auth: stringField(),
            username: stringField(),
          }),
        },
      },
    },
  ).handler(
    async () => ({
      headers: {
        one: 'foo',
        many: [42],
        cookies: {
          auth: 'Payload',
          username: 'User',
        },
      },
    })
  ),
]})

let oldLog: any = null

beforeEach(() => {
  oldLog = console.error
  console.error = jest.fn()
})

afterEach(() => {
  console.error = oldLog
})

test('if Accept media type is */* - default media type is used', async () => {
  const resp = await request(server).get('/items').set('Accept', '*/*')
  expect(resp.status).toEqual(200)
})

test('if Content-Type media type is unsupported - 415 status code is returned', async () => {
  const resp = await request(server).post('/items').set('Content-Type', 'application/xml')
  expect(resp.status).toEqual(415)
  expect(resp.text).toEqual(JSON.stringify({
    'statusCode':415,
    'isPublic':true,
    'reason': 'Not supported \'content-type\': application/xml',
  }))
})

test('if Accept media type is unsupported - 415 status code is returned', async () => {
  const resp = await request(server).get('/items').set('Accept', 'application/xml')
  expect(resp.status).toEqual(415)
  expect(resp.text).toEqual(JSON.stringify({
    'statusCode':415,
    'isPublic':true,
    'reason':'Not supported \'accept\': application/xml',
  }))
})

test('if handler fails with status code in an error - the status code is returned', async () => {
  const resp = await request(server).get('/expected-error')
  expect(resp.status).toEqual(442)
})

test('if handler fails without status code - 500 status code is returned', async () => {
  const resp = await request(server).get('/unexpected-error')
  expect(resp.status).toEqual(500)
})

test('if request is invalid - 400 status code is returned', async () => {
  const resp = await request(server).post('/items').set('Content-Type', 'application/json').send('blob')
  expect(resp.status).toEqual(400)
})

test('with-complex-headers', async () => {
  const resp = await request(server).get('/with-complex-headers')
  delete resp.headers.date
  delete resp.headers.url
  expect(resp).toMatchSnapshot()
})

test('POST', async () => {
  const resp = await request(server).post('/items').set('Content-Type', 'application/json').send({
    title: 'Title',
    description: 'Description',
  })
  expect(resp.status).toEqual(201)
  expect(resp.body).toEqual(42)
})

test('PUT', async () => {
  const resp = await request(server).put('/items/11').set('Content-Type', 'application/json').send({
    title: 'Title',
    description: 'Description',
  })
  expect(resp.status).toEqual(204)
})

test('DELETE', async () => {
  const resp = await request(server).delete('/items/11')
  expect(resp.status).toEqual(204)
})

test('PATCH', async () => {
  const resp = await request(server).patch('/items/11').set('Content-Type', 'application/json').send({
    title: 'Title-delta',
  })
  expect(resp.status).toEqual(204)
})

test('retains content-type if it is passed as a header from the handler', async () => {
  // TODO
})
