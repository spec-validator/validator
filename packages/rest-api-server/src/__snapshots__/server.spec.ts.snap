// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with-complex-headers 1`] = `
Object {
  "connection": "close",
  "content-length": "0",
  "content-type": "application/json",
  "cookies": "auth=Payload; username=User",
  "many": "42",
  "one": "foo",
}
`;
